File: Modules/Multiply.m
classdef Multiply < Module
   
    properties(SetAccess = private)
       
        output
        numberOfInputs
    end
    
    methods
        
        function this = Multiply(name, numberOfInputs)
           
            this = this@Module(name);
            this.numberOfInputs = numberOfInputs;
            
            for( n = 1:numberOfInputs )
                
                this.createInputPort();
                
                this.inputPorts(n).set( 1 );
                
            end;
            
            this.output = this.createOutputPort();
            
        end
        
        function doUpdate(this, N)
        
            y = 1;
            
            for( n = 1:this.numberOfInputs )
                
                y = y.*this.inputPorts(n).read(N);
                
            end
            
            this.output.write( y );
            
        end;
        
    end
end

File: Cpp/Modules/Source/Multiply.cpp
#include "Multiply.h"

Multiply::Multiply(size_t numberOfInputs)
{
	for (size_t n = 0; n < numberOfInputs; ++n)
	{
		CreateInputPort();
	}

	_pOutput = CreateOutputPort();
}

void Multiply::Update()
{
	auto y = 0.0;

	for (size_t n = 0; n < GetNumberOfInputs(); ++n)
	{
		y *= GetInputPort(n)->Read();
	}

	_pOutput->Write(y);
}

IOutputPort* Multiply::Output() const
{
	return _pOutput;
}
